horovod/common/ops/mpi_operations.cc

```diff
@@ -0,0 +1,405 @@
+// Copyright 2016 The TensorFlow Authors. All Rights Reserved.
+// Modifications copyright (C) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include "mpi_operations.h"
+
+namespace horovod {
+namespace common {
+
+MPI_Datatype MPIContext::GetMPIDataType(const std::shared_ptr<Tensor> tensor) {
+  return GetMPIDataType(tensor->dtype());
+}
+
+MPI_Datatype MPIContext::GetMPIDataType(const DataType dtype) {
+  switch (dtype) {
+    case HOROVOD_UINT8:
+      return MPI_UINT8_T;
+    case HOROVOD_INT8:
+      return MPI_INT8_T;
+    case HOROVOD_UINT16:
+      return MPI_UINT16_T;
+    case HOROVOD_INT16:
+      return MPI_INT16_T;
+    case HOROVOD_INT32:
+      return MPI_INT32_T;
+    case HOROVOD_INT64:
+      return MPI_INT64_T;
+    case HOROVOD_FLOAT16:
+      return mpi_float16_t;
+    case HOROVOD_FLOAT32:
+      return MPI_FLOAT;
+    case HOROVOD_FLOAT64:
+      return MPI_DOUBLE;
+    case HOROVOD_BOOL:
+      return MPI_C_BOOL;
+    case HOROVOD_BYTE:
+      return MPI_BYTE;
+    case HOROVOD_NULL:
+      return MPI_DATATYPE_NULL;
+    default:
+      throw std::logic_error("Type " + DataType_Name(dtype) +
+                             " is not supported in MPI mode.");
+  }
+}
+
+MPI_Comm MPIContext::GetMPICommunicator(Communicator comm) {
+  switch (comm) {
+    case GLOBAL:
+      return mpi_comm;
+    case LOCAL:
+      return local_comm;
+    case CROSS:
+      return cross_comm;
+    default:
+      throw std::logic_error("Communicator " + CommunicatorName(comm) +
+                             " is not supported in MPI mode.");
+  }
+}
+
+void MPI_Allreduce(MPIContext& ctx, const void* buffer_data, int64_t num_elements,
+                   TensorTableEntry& first_entry, const void* sendbuff,
+                   Communicator comm) {
+  int op = MPI_Allreduce(sendbuff != nullptr ? sendbuff : MPI_IN_PLACE, (void*) buffer_data,
+                         (int) num_elements,
+                         ctx.GetMPIDataType(first_entry.tensor),
+                         first_entry.tensor->dtype() == HOROVOD_FLOAT16 ? ctx.mpi_float16_sum : MPI_SUM,
+                         ctx.GetMPICommunicator(comm));
+  if (op != MPI_SUCCESS) {
+    throw std::logic_error("MPI_Allreduce failed, see MPI output for details.");
+  }
+}
+
+void MPI_Allgatherv(MPIContext& ctx, const void* sendbuf, int sendcount, DataType sendtype,
```

Comment:
Add to utilities files instead, instead of *_operations.cc/h

Reply:
On a related note, I find the MPI wrappers in this file to be a bit confusing generally. Are these wrappers intended to be used only in the files defining the various collective operation implementations, and not intended to be used to wrap MPI elsewhere (i.e. for coordination in `operations.cc`)?  Additionally, leaving the wrapper names unmodified from the existing unwrapped MPI API calls leads to some confusion as well. 

Could you clear up what the intent/expected use of these wrappers is?

Reply:
I agree about the confusion related the the MPI wrappers. What is the motivation for the wrappers, is it to clean up the op implementation from the MPI error handling logic? Or is it simply to standardize access to the MPI communicators through `MPIContext`?

I think if the wrappers are used, the exposed API should be identical to the underlying core MPI API (with maybe a standard modification on it, something like `MPI_Bcast_(...)` instead of `MPI_Bcast(...)` ), to make things more consistent.

Reply:
I went ahead and removed the wrappers.

For context, the wrappers were originally added as an abstraction above MPI, in case we wanted to sub in another framework that supports similar collective primitives (for NCCL).  However, the design evolved so that NCCL became coupled to MPI for now, so the wrappers no longer have much use beyond the convenience of not having to convert an error op code into an exception.

Given that they provide little benefit in the current design, and that most devs working in this codebase are accustomed to using MPI functions directly, I went ahead and removed them.